File: app/src/main/java/com/ne1c/gitteroid/models/data/RoomModel.java
package com.ne1c.gitteroid.models.data;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;
import java.util.Comparator;

public class RoomModel implements Parcelable {
    public String id;
    public String name;
    public String topic;
    public boolean oneToOne;
    public ArrayList<UserModel> users = new ArrayList<>();
    public int unreadItems;
    public int userCount;
    public int mentions;
    public String lastAccessTime;
    public boolean lurk;
    public String url;
    public String githubType;
    public boolean hide;
    public int v;
    public int listPosition = -1;


    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.id);
        dest.writeString(this.name);
        dest.writeString(this.topic);
        dest.writeByte(oneToOne ? (byte) 1 : (byte) 0);
        dest.writeList(this.users);
        dest.writeInt(this.unreadItems);
        dest.writeInt(this.userCount);
        dest.writeInt(this.mentions);
        dest.writeString(this.lastAccessTime);
        dest.writeByte(lurk ? (byte) 1 : (byte) 0);
        dest.writeString(this.url);
        dest.writeString(this.githubType);
        dest.writeInt(this.v);
    }

    public RoomModel() {
    }

    protected RoomModel(Parcel in) {
        this.id = in.readString();
        this.name = in.readString();
        this.topic = in.readString();
        this.oneToOne = in.readByte() != 0;
        this.users = new ArrayList<UserModel>();
        in.readList(this.users, ArrayList.class.getClassLoader());
        this.unreadItems = in.readInt();
        this.userCount = in.readInt();
        this.mentions = in.readInt();
        this.lastAccessTime = in.readString();
        this.lurk = in.readByte() != 0;
        this.url = in.readString();
        this.githubType = in.readString();
        this.v = in.readInt();
    }

    // r1 - old list, r2 - new list
    // copy position, status hideRoom from r1 to r2
    public static ArrayList<RoomModel> margeRooms(ArrayList<RoomModel> r1, ArrayList<RoomModel> r2) {
        for (RoomModel r1Model : r1) {
            for (RoomModel r2Model : r2) {
                if (r1Model.id.equals(r2Model.id)) {
                    r2Model.hide = r1Model.hide;
                    r2Model.listPosition = r1Model.listPosition;
                }
            }
        }

        return r2;
    }

    public static final Parcelable.Creator<RoomModel> CREATOR = new Parcelable.Creator<RoomModel>() {
        public RoomModel createFromParcel(Parcel source) {
            return new RoomModel(source);
        }

        public RoomModel[] newArray(int size) {
            return new RoomModel[size];
        }
    };

    public final static class SortedByName implements Comparator<RoomModel> {
        @Override
        public int compare(RoomModel r1, RoomModel r2) {
            return r1.name.compareTo(r2.name);
        }
    }

    public final static class SortedByPosition implements Comparator<RoomModel> {
        @Override
        public int compare(RoomModel r1, RoomModel r2) {
            if (r1.listPosition > r2.listPosition) {
                return 1;
            } else if (r1.listPosition < r2.listPosition) {
                return -1;
            } else {
                return 0;
            }
        }
    }
}

File: app/src/main/kotlin/com/ne1c/gitteroid/models/data/RoomModel.kt
package com.ne1c.gitteroid.models.data

import android.os.Parcel
import android.os.Parcelable

import java.util.*

open class RoomModel(
        var id: String = "",
        var name: String = "",
        var topic: String = "",
        var oneToOne: Boolean = false,
        var users: ArrayList<UserModel> = ArrayList(),
        var unreadItems: Int = 0,
        var userCount: Int = 0,
        var mentions: Int = 0,
        var lastAccessTime: String = "",
        var lurk: Boolean = false,
        var url: String = "",
        var githubType: String = "",
        var hide: Boolean = false,
        var v: Int = 0,
        var listPosition: Int = -1) : Parcelable {

    class SortedByName : Comparator<RoomModel> {
        override fun compare(r1: RoomModel, r2: RoomModel): Int {
            return r1.name.compareTo(r2.name)
        }
    }

    class SortedByPosition : Comparator<RoomModel> {
        override fun compare(r1: RoomModel, r2: RoomModel): Int {
            if (r1.listPosition > r2.listPosition) {
                return 1
            } else if (r1.listPosition < r2.listPosition) {
                return -1
            } else {
                return 0
            }
        }
    }

    constructor(source: Parcel) : this(source.readString(),
            source.readString(),
            source.readString(),
            1.equals(source.readInt()),
            source.createTypedArrayList(UserModel.CREATOR),
            source.readInt(),
            source.readInt(),
            source.readInt(),
            source.readString(),
            1.equals(source.readInt()),
            source.readString(),
            source.readString(),
            1.equals(source.readInt()),
            source.readInt(),
            source.readInt())

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel?, flags: Int) {
        dest?.writeString(id)
        dest?.writeString(name)
        dest?.writeString(topic)
        dest?.writeInt((if (oneToOne) 1 else 0))
        dest?.writeTypedList(users)
        dest?.writeInt(unreadItems)
        dest?.writeInt(userCount)
        dest?.writeInt(mentions)
        dest?.writeString(lastAccessTime)
        dest?.writeInt((if (lurk) 1 else 0))
        dest?.writeString(url)
        dest?.writeString(githubType)
        dest?.writeInt((if (hide) 1 else 0))
        dest?.writeInt(v)
        dest?.writeInt(listPosition)
    }

    companion object {
        @JvmField val CREATOR: Parcelable.Creator<RoomModel> = object : Parcelable.Creator<RoomModel> {
            override fun createFromParcel(source: Parcel): RoomModel = RoomModel(source)
            override fun newArray(size: Int): Array<RoomModel?> = arrayOfNulls(size)
        }
    }
}
